libraries/type-length-value/src/state.rs

```diff
@@ -300,37 +357,87 @@ impl<'data> TlvStateMut<'data> {
         let extension_ref = pod_from_bytes_mut::<V>(buffer)?;
         *extension_ref = V::default();
+        Ok((extension_ref, entry_number))
+    }
+
+    /// Packs the default TLV data into the first open slot in the data buffer.
+    /// Does not check for duplicates. Will add a new entry to the next open
+    /// slot provided there is enough space.
+    pub fn init_value<V: SplDiscriminate + Pod + Default>(
+        &mut self,
+    ) -> Result<(&mut V, usize), ProgramError> {
+        self.init_value_at_entry::<V>(None)
+    }
+
+    /// Packs the default TLV data into the first open slot in the data buffer.
+    /// If extension is already found in the buffer, it returns an error.
+    pub fn init_value_unique<V: SplDiscriminate + Pod + Default>(
+        &mut self,
+    ) -> Result<&mut V, ProgramError> {
+        let (extension_ref, _) = self.init_value_at_entry::<V>(Some(0))?;
         Ok(extension_ref)
     }
 
-    /// Packs a variable-length value into its appropriate data segment. Assumes that
-    /// space has already been allocated for the given type
-    pub fn pack_variable_len_value<V: SplDiscriminate + VariableLenPack>(
+    /// Packs a variable-length value into its appropriate data segment. Assumes
+    /// that space has already been allocated for the given type
+    fn pack_variable_len_value_at_entry<V: SplDiscriminate + VariableLenPack>(
         &mut self,
         value: &V,
+        entry_number: usize,
     ) -> Result<(), ProgramError> {
-        let data = self.get_bytes_mut::<V>()?;
+        let data = self.get_bytes_mut::<V>(entry_number)?;
         // NOTE: Do *not* use `pack`, since the length check will cause
         // reallocations to smaller sizes to fail
         value.pack_into_slice(data)
     }
 
+    /// Packs a variable-length value into its appropriate data segment, where
+    /// repeating discriminators _are_ allowed
+    pub fn pack_variable_len_value<V: SplDiscriminate + VariableLenPack>(
+        &mut self,
+        value: &V,
+        entry_number: usize,
+    ) -> Result<(), ProgramError> {
+        self.pack_variable_len_value_at_entry::<V>(value, entry_number)
+    }
+
+    /// Packs a variable-length value into its appropriate data segment, where
+    /// no repeating discriminators are allowed
+    pub fn pack_variable_len_value_unique<V: SplDiscriminate + VariableLenPack>(
+        &mut self,
+        value: &V,
+    ) -> Result<(), ProgramError> {
+        self.pack_variable_len_value_at_entry::<V>(value, 0)
+    }
+
     /// Allocate the given number of bytes for the given SplDiscriminate
-    pub fn alloc<V: SplDiscriminate>(&mut self, length: usize) -> Result<&mut [u8], ProgramError> {
-        let TlvIndices {
-            type_start,
-            length_start,
-            value_start,
-        } = get_indices(self.data, V::SPL_DISCRIMINATOR, true)?;
+    fn alloc<V: SplDiscriminate>(
+        &mut self,
+        length: usize,
+        entry_number: Option<usize>,
```

Comment:
This parameter is a bit confusing -- you only ever use `None` or `Some(0)`, so how about just passing a bool flag like `allow_repetition` and then set `entry_number` based on that?